src/patchkernel/patch_kernel_parallel.tpp

```diff
@@ -0,0 +1,85 @@
+/*---------------------------------------------------------------------------*\
+ *
+ *  bitpit
+ *
+ *  Copyright (C) 2015-2022 OPTIMAD engineering Srl
+ *
+ *  -------------------------------------------------------------------------
+ *  License
+ *  This file is part of bitpit.
+ *
+ *  bitpit is free software: you can redistribute it and/or modify it
+ *  under the terms of the GNU Lesser General Public License v3 (LGPL)
+ *  as published by the Free Software Foundation.
+ *
+ *  bitpit is distributed in the hope that it will be useful, but WITHOUT
+ *  ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ *  FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
+ *  License for more details.
+ *
+ *  You should have received a copy of the GNU Lesser General Public License
+ *  along with bitpit. If not, see <http://www.gnu.org/licenses/>.
+ *
+\*---------------------------------------------------------------------------*/
+#if BITPIT_ENABLE_MPI==1
+
+namespace bitpit {
+
+/**
+ * Confirm that the specified halo layer is the halo layer of the given cell.
+ *
+ * The check is performed looping the neighbours and searching a negihbour that belongs
+ * to the previous layer. If that neguhbour is found, the halo layer is confirmed.
+ *
+ * \param cell is the cell that will be check
+ * \param haloLayer is the halo layer the cell supposedly belongs to
+ * \param excludeList is a list of cells that will not be considered when processing the
+ * neighbours
+ * \result Return true if the cell belongs to the specified halo layer, false otherwise.
+ */
+template<typename ExcludeList>
+bool PatchKernel::confirmCellHaloLayer(const Cell &cell, int haloLayer, const ExcludeList &excludeList) const
+{
+	// Search if a face neighbour is in the previous halo layer
+	//
+	// We first search in the face neighbours only, because its faster than
+	// searching among all the neighbours.
+	//
+	// Searching for neighbours in the previous layer works also for ghosts in
```

Comment:
What about a ghost that is in the first layer but it's an edge/vertex-only neighbor of inner cells (the same for computeCellHaloLayer method)?
If I'm not wrong in the first layer of ghosts also the ghost-cells that are neighbors of the locale ones only through a vertex or an edge (no faces) are included. 
Maybe, in order to be consistent also the edge and vertex neighbors should be considered during the propagation algorithms to compute the n-layers halo (i.e. processCellsNeighbours method should work not only for face neighbors)?

Reply:
The function confirmCellHaloLayer will first check among the face vertices, if a neighbour that belongs to the previous layer is not found there, it will extend the search also to node and edge neighoburs. The first check is much faster than the latter and will cover most of the cases, therefore I thought it was a good idea to perform the "extended" check only when needed.

Reply:
Sorry, it was commented futher below in the code.